### PR TITLE
Ot riscv compliance

### PR DESCRIPTION
**Pull request by GregAC**
If you want to check out my riscv-compliance changes you may prefer to look at the riscv-compliance branch here: https://github.com/lowRISC/riscv-compliance/tree/ot (which the patches are generated from).

In addition to getting the riscv-compliance tests working again this brings us up to date with the latest version and runs more test suites in CI as well as including a couple of tests that were previously broken (ECALL and EBREAK).

### CODE REVIEW

File: sw/vendor/riscv_compliance.vendor.hjson [hunk 1]
```diff
@@ -34,5 +34,6 @@
     "riscv-test-suite/rv32ui",
     "riscv-test-suite/rv64i",
     "riscv-test-suite/rv64im",
+    "riscv-test-suite/rv32Zifencei",
```
[the remark this answers is not in the document]
Comment (GregAC) [reply sw/vendor/riscv_compliance.vendor.hjson:38]: It writes to flash memory so fails, I've added some comments to explain this.

**Comment by GregAC** [comment]
I've put another commit on this that takes out some CI tests so I can more quickly debug the CI failure in riscv-compliance.

**Comment by GregAC** [comment]
Ok this should be good to go.

Sadly the compliance suite takes a while to run in CI (~25 minutes), there's a couple of things we could do to speed it up:

1. Run the suites in parallel
2. Use the work being done for DV to allow faster printing (the tests themselves are very quick, it's the boot and the signature output that's slow).

**Comment by GregAC** [comment]
Test suite actually took 33 minutes in this example

**Comment by gkelly** [comment]
> Basing my review off https://github.com/lowRISC/riscv-compliance/compare/ot (which shows the diff), I'm happy with this approach (of saving registers before jumping into the compliance tests).
> 
> I'm worried the compliance Makefile for OT and linker script may end up being quite fragile, but I'm not sure there's a way around that that's nicer than what we have here.
> 
> There are some things I would like to see improved in follow-up patches:
> 
> 1. Parallelise the CI runs, and maybe gate them on hardware changes
> 2. Potentially move to the DV logging system which should help with speed.

Agreed about triggering on only hardware changes and looking at the DV logging system, although do these tests even log that much? I think this is probably better named `riscv_compliance_support`, but I'm happy to merge this to get it going and fix later.

File: azure-pipelines.yml [hunk 2]
```diff
@@ -235,24 +235,33 @@ jobs:
...
+        --release-version="${TOOLCHAIN_VERSION}" \
+        --update
+    displayName: Install toolchain
+  - bash: |
+      set -e
       . util/build_consts.sh
       export TARGET_SIM="$BIN_DIR/hw/top_earlgrey/Vtop_earlgrey_verilator"
       export RISCV_DEVICE=rv32imc
       export RISCV_TARGET=opentitan
+      export OT_BIN=$BIN_DIR
```
Comment (imphil) [inline azure-pipelines.yml:255]:
```suggestion
      export OT_BIN="$BIN_DIR"
```

**Comment by GregAC** [comment]
> linker script may end up being quite fragile

It actually uses the flash_link.ld from the SW tree. I forgot to remove the linker script in the OT riscv compliance target.

As it looks like the current compliance suite is maintenance only now as there's a replacement coming best we don't invest too much time in getting it running. So it is a little fragile but it should do the job.

**Pull request merged**